cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsProcessor.java

```diff
@@ -141,44 +142,46 @@ private int addPartitionMetricSamples(Cluster cluster,
-   * @return the number of skipped brokers.
+   * @param cluster The Kafka cluster
+   * @param brokerMetricSamples The set to add the broker samples to.
+   * @return The number of skipped brokers.
    */
   private int addBrokerMetricSamples(Cluster cluster,
-                                     Set<BrokerMetricSample> brokerMetricSamples) {
+                                     Set<BrokerMetricSample> brokerMetricSamples) throws UnknownVersionException {
     int skippedBroker = 0;
     MetricDef brokerMetricDef = KafkaMetricDef.brokerMetricDef();
     for (Node node : cluster.nodes()) {
       BrokerLoad brokerLoad = _brokerLoad.get(node.id());
       if (brokerLoad == null) {
         LOG.warn("Skip generating broker metric sample for broker {} because all broker metrics are missing.", node.id());
         continue;
-      } else if (!brokerLoad.allBrokerMetricsAvailable()) {
-        if (brokerLoad.missingBrokerMetrics().size() == 0) {
+      } else if (!brokerLoad.minRequiredBrokerMetricsAvailable()) {
+        if (brokerLoad.missingBrokerMetricsInMinSupportedVersion().size() == 0) {
           LOG.warn("Skip generating broker metric sample for broker {} because there are not enough topic metrics to "
                   + "generate broker metrics.", node.id());
         } else {
-          LOG.warn("Skip generating broker metric sample for broker {} because the following metrics are missing {}.",
-              node.id(), brokerLoad.missingBrokerMetrics());
+          LOG.warn("Skip generating broker metric sample for broker {} because the following required metrics are missing {}.",
+              node.id(), brokerLoad.missingBrokerMetricsInMinSupportedVersion());
         }
         continue;
       }
 
       boolean validSample = true;
-      BrokerMetricSample brokerMetricSample = new BrokerMetricSample(node.host(), node.id());
-      for (RawMetricType rawBrokerMetricType : RawMetricType.brokerMetricTypes()) {
-        // We require the broker to report all the metric types. Otherwise we skip the broker.
-        if (!brokerLoad.brokerMetricAvailable(rawBrokerMetricType)) {
-          skippedBroker++;
-          validSample = false;
-          LOG.warn("Skip generating broker metric sample for broker {} because it does not have %s metrics or "
-                        + "the metrics are inconsistent.", node.id(), rawBrokerMetricType);
-          break;
-        } else {
-          MetricInfo metricInfo = brokerMetricDef.metricInfo(KafkaMetricDef.forRawMetricType(rawBrokerMetricType).name());
-          double metricValue = brokerLoad.brokerMetric(rawBrokerMetricType);
-          brokerMetricSample.record(metricInfo, metricValue);
+      BrokerMetricSample brokerMetricSample = new BrokerMetricSample(node.host(), node.id(), brokerLoad._brokerSampleDeserializationVersion);
+      for (Map.Entry<Byte, Set<RawMetricType>> entry : RawMetricType.brokerMetricTypesDiffByVersion().entrySet()) {
```

Comment:
If we reach here, it means `setNullableBrokerMetrics()` is already executed and it already fill all the nullable field, so this check will always pass. is my understanding correct?

Reply:
Correct -- as I commented on a previous feedback, this sanity check ensures that each metric either has a value or has correctly been nullified.